archive/extract_activity.go

```diff
@@ -0,0 +1,177 @@
+package archive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/safeopen"
+	"github.com/mholt/archiver/v4"
+	"go.artefactual.dev/tools/temporal"
+)
+
+const ExtractActivityName = "extract-archive-activity"
+
+var (
+	ErrNotAFile       = errors.New("Not a file")
+	ErrInvalidArchive = errors.New("Invalid archive")
+)
+
+type ExtractActivity struct {
+	cfg Config
+}
+
+func NewExtractActivity(cfg Config) *ExtractActivity {
+	return &ExtractActivity{cfg: cfg}
+}
+
+type ExtractActivityParams struct {
+	// SourcePath is the path of the archive to be extracted
+	SourcePath string
+}
+
+type ExtractActivityResult struct {
+	// DestPath is the path of the extracted archive contents.
+	DestPath string
+}
+
+// Execute extracts the content of SourcePath to a temporary directory, then
+// returns the DestPath of the extracted files.
+//
+// If SourcePath is a directory an ErrNotAFile error is returned. If SourcePath
+// is a file, but not a valid archive, an ErrInvalidArchive error is returned.
+func (a *ExtractActivity) Execute(
+	ctx context.Context, params *ExtractActivityParams,
+) (*ExtractActivityResult, error) {
+	logger := temporal.GetLogger(ctx)
+	logger.V(1).Info("Executing ExtractActivity", "SourcePath", params.SourcePath)
+
+	a.cfg.setDefaults()
+
+	dest, err := a.extract(ctx, params.SourcePath)
+	if err != nil {
+		switch err {
+		case ErrNotAFile:
+			logger.V(2).Info("ExtractActivity: not a file", "SourcePath", params.SourcePath)
+			return nil, err
+		case ErrInvalidArchive:
+			logger.V(2).Info("ExtractActivity: not a valid archive", "SourcePath", params.SourcePath)
+			return nil, err
+		default:
+			return nil, fmt.Errorf("extract archive: %v", err)
+		}
+	}
+
+	dest, err = skipTopLevelDir(dest)
+	if err != nil {
+		return nil, fmt.Errorf("extract archive: skipTopLevelDir: %v", err)
+	}
+
+	return &ExtractActivityResult{DestPath: dest}, nil
+}
+
+func (a *ExtractActivity) extract(ctx context.Context, src string) (string, error) {
+	fi, err := os.Stat(src)
+	if err != nil {
+		return "", err
+	}
+
+	if fi.IsDir() {
+		return src, ErrNotAFile
+	}
+
+	f, err := os.Open(src) // #nosec G304 -- trusted path.
```

Comment:
We could use `filepath.Clean(src)` and remove the comment.

Reply:
I've seen comments online that `filepath.Clean()` prevents path traversal, but I don't think it does: https://go.dev/play/p/RSUJlM4aM_q

I thought about using https://github.com/google/safeopen, e.g.:

```go
fd, err := safeopen.OpenBeneath(basePath, src)
```

But I don't know what `basePath` would be.